sqlglot/parser.py

```diff
@@ -1346,6 +1346,32 @@ def _parse_command(self) -> exp.Command:
             exp.Command, this=self._prev.text.upper(), expression=self._parse_string()
         )
 
+    def _try_parse(
+        self, parse_method: t.Callable, retreat: bool = False
+    ) -> t.Optional[exp.Expression]:
+        """
+        Attemps to backtrack if a parse function that contains a try/catch internally raises an error. This behavior can
+        be different depending on the uset-set ErrorLevel, so _try_parse aims to solve this by setting & resetting
+        the parser state accordingly
+        """
+        index = self._index
+        errors = self.errors.copy()
+        error_level = self.error_level
+
+        self.error_level = ErrorLevel.IMMEDIATE
+        this = None
+        try:
+            this = parse_method()
+        except ParseError:
+            this = None
+        finally:
+            if not this or retreat:
```

Comment:
Why would we ever _not_ want to retreat? Isn't the whole point here that `parse_method` could consume input that we need to successfully take an alternative parser path?

Reply:
This would be the case that `parse_method ` is on the right path and does not need to backtrack; in the Clickhouse case, `parse_method` is `None` if the CTE is expression-based, but if it's a typical CTE then `parse_method` contains the actual sub-tree, so retreating here wouldn't be correct right?

Reply:
Consuming input does not necessarily mean it's on the right path. We could consume input and then fail. I think the disconnect is that we should _always_ retreat, but only do it in the `except` branch.